dvc/tree/dvc.py

```diff
@@ -11,6 +12,24 @@
 logger = logging.getLogger(__name__)
 
 
+# cache metadata for sequential exists/isdir/isfile/etc calls
+@lru_cache(maxsize=1)
+def _get_metadata(tree, path_info):
```

Comment:
same `lru_cache(maxsize=1)` optimization that we use in git tree `_get_object_by_path`

Reply:
With git it is pretty safe, as tree is read only, tied to a specific reference. Are there scenarios in which this could cause a bug?

Reply:
Btw, which method was causing too much `metadata` calls? Would it be solved by our planned exception unification and moving to an exception-based workflow?

Reply:
In `RepoTree.get_hash` we end up making several repeated `metadata` calls through `DvcTree.exists`. I think moving to an exception based workflow and avoiding all the `exists` calls would definitely help in this case.

Reply:
I suppose in theory we could run into issues here with the dvc repo (and repo tree) being modified in between metadata calls. I can adjust this so that we only cache metadata calls within the scope of `dvctree.walk`, since the tree should not be changing within a single `walk` call.

Reply:
Ended up reverting this change, as the main performance blocker in our metadata calls was still `_get_granular_checksum`. As noted before though, moving to an exception based workflow should cut down on duplicated metadata calls